test/type_check_test.rb

```diff
@@ -1507,4 +1507,30 @@ def test_lambda__hint_is_untyped
       YAML
     )
   end
+
+  def test_safe_navigation_operator__or_hint
+    run_type_check_test(
+      signatures: {
+        "a.rbs" => <<~RBS
+          class Hello
+            def foo: (Integer?) -> Integer
+          end
+        RBS
+      },
+      code: {
+        "a.rb" => <<~RUBY
+          class Hello
+            def foo(a)
+              a&.then {|x| x.infinite? } || -1
```

Comment:
Steep reported a confusing type error: 

```
Cannot find compatible overloading of method `then` of type `::Integer`
```

It was because the hint given to `&.then` call is `Integer`, and it is forwarded to block typing where `x.infinite?` has `Integer?`.